File: 라이 대회알고리즘/day11. 백트래킹(Backtracking)/9663_N-Queen.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func step1(N int, queensX []int, queensY []int, row int) (result int) {
	if row == N {
		if len(queensX) == N {
			return 1
		} else {
			return 0
		}
	}

	for col := 0; col < N; col++ {
		nx, ny := row, col
		valid := true
		for i := 0; i < len(queensX); i++ {
			queenX, queenY := queensX[i], queensY[i]
			if queenY == ny || queenX+queenY == nx+ny || queenX-queenY == nx-ny {
				valid = false
				break
			}
		}
		if valid {
			result += step1(N, append(queensX, nx), append(queensY, ny), row+1)
		}
	}
	return result
}

func main() {

	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)

	var N int
	fmt.Fscanln(reader, &N)

	fmt.Fprintln(writer, step1(N, []int{}, []int{}, 0))
	writer.Flush()
}

File: 라이 대회알고리즘/day11. 백트래킹(Backtracking)/9663_N-Queen.py
import sys

sys.setrecursionlimit(10 ** 6)
readline = sys.stdin.readline

direction = [
    (0, 1),
    # (0, -1),
    (1, 0),
    # (-1, 0)
]


def step1(N, queens, row):
    if row == N and len(queens) == N:
        return 1

    answer = 0
    for col in range(N):
        nx, ny = row, col

        valid = True
        for queen_x, queen_y in queens:
            if queen_y == ny or queen_x + queen_y == nx + ny or queen_x - queen_y == nx - ny:
                valid = False
                break
        if valid:
            queens.append((nx, ny))
            answer += step1(N, queens, row + 1)
            queens.pop()

    return answer


def step2(N):
    q = set()
    q.add((0, ''))
    answer = 0
    while len(q) != 0:
        row, queens = q.pop()
        queens = queens.split('|') if len(queens) != 0 else []
        if len(queens) == N:
            answer += 1
            continue

        for col in range(N):
            nx, ny = row, col
            col_valid = True
            for queen in queens:
                queen_x, queen_y = map(int, queen.split(","))
                if queen_y == ny or queen_x + queen_y == nx + ny or queen_x - queen_y == nx - ny:
                    col_valid = False
                    break
            if col_valid:
                queens.append(f'{nx},{ny}')
                q.add((row+1, '|'.join(queens)))
                queens.pop()
    return answer


if __name__ == '__main__':
    N = int(readline())

    print(step1(N, [], 0))
    # print(step2(N))
